File: smoke_tests/swift/csrf.swift
// Cross-Site Request Forgery (CSRF) vulnerabilities in Swift
import Foundation

class CsrfVulnerabilities {

    // Test 1: State-changing GET request
    func handleGetRequest(action: String, userId: Int) {
        // VULNERABLE: State change via GET
        if action == "delete" {
            deleteUser(userId: userId)
        } else if action == "promote" {
            promoteUser(userId: userId)
        }
    }

    // Test 2: Missing CSRF token validation
    func handlePostRequest(params: [String: Any]) {
        // VULNERABLE: No CSRF token check
        if let action = params["action"] as? String {
            performAction(action)
        }
    }

    // Test 3: CSRF token in URL
    func generateFormUrl(token: String) -> String {
        // VULNERABLE: Token exposed in URL (visible in logs, referrer)
        return "https://api.example.com/action?csrf_token=\(token)"
    }

    // Test 4: Predictable CSRF token
    func generateCsrfToken(userId: Int) -> String {
        // VULNERABLE: Predictable token
        return "csrf_\(userId)_\(Date().timeIntervalSince1970)"
    }

    // Test 5: CSRF token without session binding
    func validateCsrfToken(token: String) -> Bool {
        // VULNERABLE: Not bound to user session
        return validTokens.contains(token)
    }

    // Test 6: SameSite cookie not set
    func setCookie(response: inout HTTPURLResponse, token: String) {
        // VULNERABLE: Missing SameSite attribute
        let cookie = "session=\(token); Path=/; HttpOnly"
        // Should have SameSite=Strict or SameSite=Lax
    }

    // Test 7: CORS misconfiguration enabling CSRF
    func handleCorsRequest(origin: String) -> [String: String] {
        // VULNERABLE: Reflecting any origin
        return [
            "Access-Control-Allow-Origin": origin,
            "Access-Control-Allow-Credentials": "true"
        ]
    }

    // Test 8: JSON endpoint without CSRF
    func handleJsonPost(data: Data) -> [String: Any] {
        // VULNERABLE: JSON POST without CSRF token
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return ["error": "Invalid JSON"]
        }
        return processData(json)
    }

    // Test 9: Form submission without token
    func renderForm() -> String {
        // VULNERABLE: No CSRF token in form
        return """
        <form action="/transfer" method="POST">
            <input name="amount" type="text">
            <input name="recipient" type="text">
            <button type="submit">Transfer</button>
        </form>
        """
    }

    // Test 10: CSRF token reuse
    func getCsrfToken(sessionId: String) -> String {
        // VULNERABLE: Same token reused across requests
        if let existing = tokenCache[sessionId] {
            return existing
        }
        let token = UUID().uuidString
        tokenCache[sessionId] = token
        return token
    }

    // Test 11: Weak CSRF validation
    func validateRequest(headers: [String: String], params: [String: String]) -> Bool {
        // VULNERABLE: Only checking referer, which can be spoofed
        guard let referer = headers["Referer"] else { return false }
        return referer.contains("example.com")
    }

    // Test 12: API without CSRF for web clients
    func handleApiCall(method: String, path: String, headers: [String: String]) {
        // VULNERABLE: No CSRF protection for web API calls
        if method == "POST" || method == "PUT" || method == "DELETE" {
            // Should validate CSRF token for web clients
            processApiRequest(path: path)
        }
    }

    private var validTokens: Set<String> = []
    private var tokenCache: [String: String] = [:]

    private func deleteUser(userId: Int) {}
    private func promoteUser(userId: Int) {}
    private func performAction(_ action: String) {}
    private func processData(_ data: [String: Any]) -> [String: Any] { [:] }
    private func processApiRequest(path: String) {}
}

File: smoke_tests/php/csrf.php
<?php
// Cross-Site Request Forgery (CSRF) vulnerabilities in PHP

// Test 1: No CSRF token on form
function transfer_funds() {
    // VULNERABLE: No CSRF token validation
    $amount = $_POST['amount'];
    $to_account = $_POST['to_account'];
    process_transfer($amount, $to_account);
}

// Test 2: State change via GET
function delete_account() {
    $id = $_GET['id'];
    // VULNERABLE: DELETE via GET request
    delete_user($id);
    header('Location: /');
}

// Test 3: Predictable token
function get_csrf_token() {
    // VULNERABLE: Predictable token
    return md5($_SESSION['user_id'] . 'secret');
}

// Test 4: Token validation bypass
function validate_csrf() {
    $token = $_POST['csrf_token'];
    // VULNERABLE: Empty token accepted
    if (empty($token) || $token === $_SESSION['csrf_token']) {
        return true;
    }
    return false;
}

// Test 5: Token not tied to session
function generate_weak_token() {
    // VULNERABLE: Token not tied to session
    return bin2hex(random_bytes(16));
    // Should be: hash('sha256', session_id() . random_bytes(16))
}

// Test 6: CORS misconfiguration allowing CSRF
function cors_endpoint() {
    // VULNERABLE: CORS allows any origin with credentials
    header('Access-Control-Allow-Origin: *');
    header('Access-Control-Allow-Credentials: true');
    echo json_encode(['data' => 'sensitive']);
}

// Test 7: Cookie without SameSite
function set_auth_cookie() {
    // VULNERABLE: No SameSite attribute (pre PHP 7.3 style)
    setcookie('auth', 'value', time() + 3600, '/', '', true, true);
    // Missing SameSite
}

// Test 8: Admin action without CSRF
function promote_user() {
    // VULNERABLE: Admin action without protection
    $user_id = $_POST['user_id'];
    set_user_role($user_id, 'admin');
}

// Test 9: Token in query string
function process_payment() {
    $token = $_GET['csrf_token'];
    // VULNERABLE: Token in URL can leak via Referer
    if (validate_token($token)) {
        process_payment_internal();
    }
}

// Test 10: JSON endpoint without CSRF
function api_update() {
    // VULNERABLE: JSON requests need CSRF too
    $data = json_decode(file_get_contents('php://input'), true);
    update_user_profile($data);
}

// Test 11: Token reuse
function reusable_token() {
    // VULNERABLE: Token should be single-use
    $token = $_POST['csrf_token'];
    if ($token === $_SESSION['csrf_token']) {
        // Process - but token remains valid
        return true;
    }
    return false;
}

// Test 12: Referrer-based CSRF protection
function referrer_check() {
    $referrer = $_SERVER['HTTP_REFERER'];
    // VULNERABLE: Referrer can be spoofed or stripped
    if (strpos($referrer, 'example.com') !== false) {
        return true;
    }
    return false;
}

// Helper functions
function process_transfer($amount, $to) {}
function delete_user($id) {}
function set_user_role($id, $role) {}
function validate_token($token) { return true; }
function process_payment_internal() {}
function update_user_profile($data) {}
?>
